modules/geochemistry/src/userobjects/NodalVoidVolume.C

```diff
@@ -0,0 +1,226 @@
+    rebuildStructures(); // reinitialize _nodal_void_volume and rebuild MPI communication lists
+}
+
+void
+NodalVoidVolume::meshChanged()
+{
+  ElementUserObject::meshChanged();
+  _rebuilding_needed = true; // signal that internal datastructures need rebuilding
+}
+
+void
+NodalVoidVolume::timestepSetup()
+{
+  ElementUserObject::timestepSetup();
+  if (_rebuilding_needed)
+    rebuildStructures(); // reinitialize _nodal_void_volume and rebuild MPI communication lists
+}
+
+void
+NodalVoidVolume::rebuildStructures()
+{
+  // Because of the RelationshipManager, this processor knows about all its elements as well as 1
+  // layer of ghost elements.  Hence, the loop over getEvaluableElementRange below goes over all the
+  // local elements as well as 1 layer of ghost elements, and all their nodes are recorded.  The
+  // ghosted elements are not visited in execute() so the nodal volume is incorrectly computed for
+  // all the nodes belonging to ghosted elements.  So MPI communication of nodal volume info is
+  // needed (implemented in exchangeGhostedInfo).
+  _nodal_void_volume.clear();
+  for (const auto & elem : _fe_problem.getEvaluableElementRange())
+    if (this->hasBlocks(elem->subdomain_id()))
+      for (unsigned i = 0; i < elem->n_nodes(); ++i)
+        if (_nodal_void_volume.count(elem->node_ptr(i)) == 0)
+          _nodal_void_volume[elem->node_ptr(i)] = 0.0;
+  if (_app.n_processors() > 1)
+    buildCommLists();
+  _rebuilding_needed = false;
+}
+
+void
+NodalVoidVolume::buildCommLists()
+{
+  // global_node_nums_to_receive[pid] is a list of node numbers connected to elements owned by
+  // processor pid.  _nodes_to_receive[pid] is a list of node pointers.  Both lists have the same
+  // ordering
+  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_receive;
+  _nodes_to_receive.clear();
+  // run through all elements known by this processor (the owned + 1 layer of ghosted elements),
+  // recording nodes that are attached to elements that aren't owned by this processor
+  for (const auto & elem : _fe_problem.getEvaluableElementRange())
+    if (this->hasBlocks(elem->subdomain_id()))
+    {
+      const processor_id_type elem_pid = elem->processor_id();
+      if (elem_pid != _my_pid)
+      {
+        if (global_node_nums_to_receive.count(elem_pid) == 0)
+        {
+          global_node_nums_to_receive[elem_pid] = std::vector<dof_id_type>();
+          _nodes_to_receive[elem_pid] = std::vector<const Node *>();
+        }
+        for (unsigned i = 0; i < elem->n_nodes(); ++i)
+          if (std::find(global_node_nums_to_receive[elem_pid].begin(),
+                        global_node_nums_to_receive[elem_pid].end(),
+                        elem->node_id(i)) == global_node_nums_to_receive[elem_pid].end())
+          {
+            global_node_nums_to_receive[elem_pid].push_back(elem->node_id(i));
+            _nodes_to_receive[elem_pid].push_back(elem->node_ptr(i));
+          }
+      }
+    }
```

Comment:
```suggestion
  std::map<processor_id_type, std::set<const Node *>> filled_nodes;
  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_receive;
  _nodes_to_receive.clear();
  // run through all elements known by this processor (the owned + 1 layer of ghosted elements),
  // recording nodes that are attached to elements that aren't owned by this processor
  for (const auto & elem : _fe_problem.getEvaluableElementRange())
    if (this->hasBlocks(elem->subdomain_id()))
    {
      const processor_id_type elem_pid = elem->processor_id();
      if (elem_pid != _my_pid)
      {
        auto & pid_nodes = filled_nodes[elem_pid];
        for (const auto & node : elem->node_ref_range())
          if (pid_nodes.insert(&node).second) // true if we haven't filled this node
          {
            global_node_nums_to_receive[elem_pid].push_back(node.id());
            _nodes_to_receive[elem_pid].push_back(&node);
          }
      }
    }
```
Adds a set to do the checking if you've filled a node already instead of doing the linear search on `global_node_nums_to_receive` This one's not as important as the other map suggested changes, but I think it cleans it up a little!

Reply:
Thanks